csi/node_server.go

```diff
@@ -197,8 +204,48 @@ func (ns *NodeServer) NodeUnstageVolume(
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
-func (ns *NodeServer) NodeGetVolumeStats(ctx context.Context, in *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "")
+func (ns *NodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+	existVol, err := ns.apiClient.Volume.ById(req.GetVolumeId())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if existVol == nil {
+		msg := fmt.Sprintf("NodeGetVolumeStats: the volume %q not exists", req.GetVolumeId())
+		logrus.Warn(msg)
+		return nil, status.Error(codes.NotFound, msg)
+	}
```

Comment:
CSI spec requires to return `codes.NotFound` if the volume does not exist. But, I am worried that this part might affect etcd performance.

Reply:
NotFound is the correct return the CO (i.e. kubernetes) is then responsible for trying again with exponential backoff.
Any specific concern?

Reply:
regard to the comment https://github.com/longhorn/longhorn-manager/pull/691#discussion_r496068884
> NotFound is the correct return the CO (i.e. kubernetes) is then responsible for trying again with exponential backoff.
Any specific concern?

My concern is that every time users scrape Kubelet metrics, we call apiClient. This might affect etcd performance when use scrapes metrics too often.

Reply:
Good thinking ahead 👍 
In this case, we are not actually calling against the kubernetes api (ETCD).

The api client is a client against the longhorn api, and will call the below method on the longhorn side:
https://github.com/longhorn/longhorn-manager/blob/93f30d51e817e17f0331908ce81f52f2b8271062/api/volume.go#L65

Which will then retrieve the volume from the datastore via the volume-manager
https://github.com/longhorn/longhorn-manager/blob/93f30d51e817e17f0331908ce81f52f2b8271062/api/volume.go#L79

Reply:
Oh. I see. Thank you so much for the explanation!